core/blockchain.go

```diff
@@ -857,6 +867,9 @@ func (self *BlockChain) InsertChain(chain types.Blocks) (int, error) {
 				continue
 			}
 
+			self.futureMu.Lock()
+			defer self.futureMu.Unlock()
```

Comment:
This is inside a for loop. It can go very wrong if you get two errors as it will deadlock the entire system.